scripts/export_environment.py

```diff
@@ -170,32 +162,67 @@ def resolve_env(env: Environment) -> Environment:
     >>> resolve_env({'x': '{y}', 'y': '42'})
     {'x': '42', 'y': '42'}
 
-    Literal curly braces need to be escaped:
+    Unmatched curly braces and curly braces that do not surround valid syntax
+    keywords will be ignored:
 
-    >>> resolve_env({'x': '{{', 'y': '}}', 'z': '{{x}}'})
-    {'x': '{', 'y': '}', 'z': '{x}'}
+    >>> resolve_env({'x': '{{', 'y': '{ z }', 'z': '{42}'})
+    {'x': '{{', 'y': '{ z }', 'z': '{42}'}
 
-    Does not support transitive references:
+    Transitive references are supported:
 
     >>> resolve_env({'x': '{y}', 'y': '{z}', 'z': '42'})
-    {'x': '{z}', 'y': '42', 'z': '42'}
+    {'x': '42', 'y': '42', 'z': '42'}
 
-    Transitive resolution would yield
+    Generative references formed during resolution are supported:
 
-    {'x': '42', 'y': '42', 'z': '42'}
+    >>> resolve_env({'x': '{y}}', 'y': '{z', 'z': '42'})
+    {'x': '42', 'y': '{z', 'z': '42'}
 
-    Does not support generative ones (references formed during resolution):
+    Consistent with a $missing_var in the shell a references to a missing key
+    yields an empty string replacement:
 
-    >>> resolve_env({'x': '{y}}}', 'y': '{{z', 'z': '42'})
-    {'x': '{{z}', 'y': '{z', 'z': '42'}
+    >>> resolve_env({'x': 'y is {y}'})
+    {'x': 'y is '}
 
-    which could be
+    A circular reference causes an exception:
 
-    {'x': '42', 'y': '{z', 'z': '42'}
+    >>> resolve_env({'x': '{y}', 'y': '{x}'})
+    Traceback (most recent call last):
+    ...
+    ValueError: Circular reference not allowed.
     """
-    # `{missing_var}` references in values should yield the empty string,
-    # consistent with $missing_var in the shell.
-    return {k: v.format_map(defaultdict(str, env)) for k, v in env.items()}
+    while True:
+        resolved_env = {k: _custom_format_map(k, v, env) for k, v in env.items()}
+        if env == resolved_env:
+            return resolved_env
+        else:
+            env = resolved_env
+
+
+def _custom_format_map(key: str, val: str, mapping: Mapping[str, str]) -> str:
```

Comment:
Not exactly an improvement because it feels like it's adding more assumptions about format_map. Also, this is complex enough to warrant dedicated doctests. Those would also help to more accurately document the behavior (the docstring below is somewhat ambiguous). For example, what is a "valid syntax keyword"? (rhetorical question)

We always have to weight the gain in functionality against the complexity of the approach and this feels off balance. I think the breadth first approach in the calling function is what gets us into trouble. I think I have an idea of how this can be simplified at an acceptable cut in functionality.